{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "bSIZ8c5Nm7zT"
   },
   "source": [
    "## Functions\n",
    "\n",
    "This notebook includes all the functiones used for visualisation and analysis. Here you can modify more specific parameters. If you need help with any of these, use help(name_of_function) on any of the analysis notebooks.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "WfCYV1ADjnzp"
   },
   "outputs": [],
   "source": [
    "# determine how many pixels (the resolution) of the plots\n",
    "plt.rcParams['figure.dpi'] = 200"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ymrDYp4sAUqG"
   },
   "source": [
    "### Create a map showing the sample distribution\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "unrJnaNSATUx"
   },
   "outputs": [],
   "source": [
    "def map_samples(pf6plus, label, stock_image=False, individual_sites=False, save_fig=False):\n",
    "    \n",
    "    ''' Plot a world map showcasing the countries where the  Pf6+ samples come from.\n",
    "    '''\n",
    "    #temp supressing warning for slicing subsets (mapping with exact number on both dataframes)\n",
    "    pd.options.mode.chained_assignment = None\n",
    "\n",
    "    fig = plt.figure()\n",
    "    ax = fig.add_subplot(1, 1, 1, projection=ccrs.PlateCarree())\n",
    "    ax.set_extent([-80, 153, -26, 29], crs=ccrs.PlateCarree())\n",
    "\n",
    "    # dictionary storing the colours for each population\n",
    "    population_colours = collections.OrderedDict()\n",
    "    population_colours['SAM'] = \"#4daf4a\"\n",
    "    population_colours['WAF'] = \"#e31a1c\"\n",
    "    population_colours['CAF'] = \"#fd8d3c\" \n",
    "    population_colours['EAF'] = \"#fecc5c\" \n",
    "    population_colours['SAS'] = \"#984ea3\"\n",
    "    population_colours['WSEA'] = \"#9ecae1\" \n",
    "    population_colours['ESEA'] = \"#3182bd\" \n",
    "    population_colours['OCE'] = \"#f781bf\"\n",
    "\n",
    "    pf6plus['population_colour'] = pf6plus['Population'].map(population_colours)\n",
    "\n",
    "    # get country borders\n",
    "    resolution = '10m'\n",
    "    category = 'cultural'\n",
    "    name = 'admin_0_countries'\n",
    "    shpfilename = shapereader.natural_earth(resolution, category, name)\n",
    "\n",
    "    # read the shapefile using geopandas\n",
    "    df = geopandas.read_file(shpfilename)\n",
    "    \n",
    "    countries = np.unique(pf6plus['Country'], return_counts=True)[0]\n",
    "    samples = np.unique(pf6plus['Country'], return_counts=True)[1]\n",
    "\n",
    "\n",
    "    # normalising using the (# of samples) corresponding to the country with the highest number of samples `max(np.unique(pf6plus['Country'], return_counts=True)[1])`.\n",
    "    samples_norm = samples/2302\n",
    "    # choose colourmap palette for map\n",
    "    cmap = plt.cm.get_cmap('YlOrRd')\n",
    "\n",
    "    #choose colourmap palette for country category scatterplot\n",
    "    #country_cat = np.unique(pf6plus['Country']\n",
    "    col = np.linspace(0, 1, len(np.unique(pf6plus['Country'])))\n",
    "    colordict = dict(zip(np.unique(pf6plus['Country']), col))    \n",
    "    pf6plus['Country_color'] = pf6plus['Country'].apply(lambda x: colordict[x])\n",
    "\n",
    "    adm1_cmap = plt.cm.get_cmap('YlOrRd')\n",
    "    for country, samp_norm in zip(countries, samples_norm):\n",
    "        # read the borders for each country\n",
    "        borders = df.loc[df['ADMIN'] == country,'geometry'].values[0]\n",
    "        \n",
    "        if individual_sites:\n",
    "          #admin sites by colour of country\n",
    "          ax.scatter(pf6plus['Longitude_adm1'], pf6plus['Latitude_adm1'], c=pf6plus['Country_color'],s=3.5, marker='o', linewidth=0.3)\n",
    "\n",
    "        # plot the coloured country on a map\n",
    "        else:\n",
    "          ax.add_geometries([borders], crs=ccrs.PlateCarree(), facecolor=cmap(samp_norm), edgecolor='none', zorder=1)\n",
    "    #add colour bar on the map\n",
    "    if individual_sites == False:\n",
    "          dummy_scat = ax.scatter([None] * len(samples), [None] * len(samples), c=samples_norm, cmap=cmap, vmin=0, vmax=2302, zorder=0)\n",
    "          fig.colorbar(mappable=dummy_scat, label='Number of samples', orientation='vertical', shrink=.3) #\n",
    "\n",
    "    #choose whether to have a stock image on the map\n",
    "    if stock_image:\n",
    "      ax.stock_img()\n",
    "    else:\n",
    "      ax.add_feature(cartopy.feature.LAND, color = 'lightgray')\n",
    "      ax.add_feature(cartopy.feature.OCEAN)\n",
    "      ax.add_feature(cartopy.feature.COASTLINE, edgecolor = 'grey')\n",
    "    plt.title(label + \" Sampled Countries\", fontsize=6)\n",
    "    plt.show()\n",
    "\n",
    "    #save figure as .png with label & data as name\n",
    "    if save_fig:\n",
    "      fig.savefig(f'{label}_{strftime(\"%Y-%m-%d\")}.png', bbox_inches='tight', dpi=200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "QfOMzT9DjUzY"
   },
   "outputs": [],
   "source": [
    "def temporal_samples(samples = 'Pf6+'):\n",
    "    \n",
    "    ''' Plot a stacked bar plot showcasing the temporal distribution of the samples in Pf6+.\n",
    "    '''\n",
    "\n",
    "    fig, ax = plt.subplots()\n",
    "\n",
    "    if (samples == 'Pf6+'):\n",
    "\n",
    "      ax.bar(np.unique(pf6plus.loc[pf6plus['Process'] == 'WGS']['Year']), np.unique(pf6plus.loc[(pf6plus['Process'] == 'WGS'), 'Year'], return_counts=True)[1], label='Pf6')\n",
    "      ax.bar(np.unique(pf6plus.loc[pf6plus['Process'] != 'WGS']['Year']), np.unique(pf6plus.loc[(pf6plus['Process'] != 'WGS'), 'Year'], return_counts=True)[1], label='GenRe-Mekong')\n",
    "      \n",
    "    min_year = (pf6plus['Year']).min()\n",
    "    max_year = (pf6plus['Year']).max()\n",
    "\n",
    "    ax.set_xlim(min_year, max_year)\n",
    "\n",
    "    ax.set_ylabel('# of samples')\n",
    "    ax.set_title('Year of Collection')\n",
    "    ax.legend()\n",
    "\n",
    "    ##add versions for individual version (TO DO)\n",
    "    # else: \n",
    "    #   ax.bar(np.unique(pf6plus.loc[pf6plus['Process'] == 'WGS']['Year']), np.unique(pf6plus.loc[(pf6plus['Process'] == 'WGS'), 'Year'], return_counts=True)[1], label='Pf6')\n",
    "\n",
    "\n",
    "    plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "HeFiSjigZJmp"
   },
   "source": [
    "###Tabulate drug resistant variants\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "yNGCmgRDh6LK"
   },
   "outputs": [],
   "source": [
    "def tabulate_drug_resistant(drug, country = None, population = None, year = None, bin=False):\n",
    "\n",
    "    ''' Tabulate the frequency of drug resistant samples per country/year \n",
    "    \n",
    "    Parameters:\n",
    "      - drug: Any of the drugs in the Pf6+ dataframe ['Artemisinin', 'Chloroquine', 'DHA-PPQ', 'Piperaquine', 'Pyrimethamine', 'S-P', 'S-P-IPTp', 'Sulfadoxine']\n",
    "      - country: Any of the countries in the Pf6+ dataframe (if specified, population value is not used) ['Bangladesh', 'Benin', 'Burkina Faso', 'Cambodia', 'Cameroon', 'Colombia', 'Congo DR', 'Ethiopia', 'Gambia', 'Ghana', 'Guinea', 'India', 'Indonesia', 'Ivory Coast', 'Kenya', 'Laos', 'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mozambique', 'Myanmar', 'Nigeria', 'Papua New Guinea', 'Peru', 'Senegal', 'Tanzania', 'Thailand', 'Uganda', 'Viet Nam']\n",
    "      - population: Any of the populations in the Pf6+ dataframe ['CAF', 'EAF', 'ESEA', 'OCE', 'SAM', 'SAS', 'WAF', 'WSEA']\n",
    "      - year: An array with the year(s) in the Pf6+ dataframe [2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]\n",
    "      - bin: If True, all the years between the specified values will be used. If False, individual years are used.\n",
    "\n",
    "    Returns:\n",
    "      A dataframe showing the number of Resistant, Sensitive & Undetermined \n",
    "      samples using the drug/country/year (or) drug/country/year combination provided. . The total number\n",
    "      of samples and drug resistant frequency is also provided. \n",
    "    '''\n",
    "\n",
    "\n",
    "    if all([country,year]):\n",
    "\n",
    "      if bin:\n",
    "        samples = pf6plus.loc[(pf6plus.Country.isin([country])) & (pf6plus.Year.isin([b for b in range(min(year),max(year))]))]\n",
    "        print(drug + ' resistant samples ' + 'in ' + str(country) + ' from ' + str(min(year)) + ' to ' + str(max(year)) )\n",
    "      else:\n",
    "        samples = pf6plus.loc[(pf6plus.Country.isin([country])) & (pf6plus.Year.isin(year))]\n",
    "        print(drug + ' resistant samples in ' + str(country) + ' on ' + str(year))\n",
    "\n",
    "      phenotypes = samples.groupby([drug]).size().fillna(0).astype(int).to_frame('Samples').transpose()\n",
    "      \n",
    "    elif all([population,year]):\n",
    "\n",
    "      if bin:\n",
    "\n",
    "        samples = pf6plus.loc[(pf6plus.Population.isin([population])) & (pf6plus.Year.isin([b for b in range(min(year),max(year))]))]\n",
    "        print(drug + ' resistant samples ' + 'in ' + str(population)  + ' from ' + str(min(year)) + ' to ' + str(max(year)) )\n",
    "      else:\n",
    "        samples = pf6plus.loc[(pf6plus.Population.isin([population])) & (pf6plus.Year.isin(year))]\n",
    "        print(drug + ' resistant samples in ' + str(population) + ' on ' + str(year))\n",
    "\n",
    "      phenotypes = samples.groupby([drug]).size().fillna(0).astype(int).to_frame('Samples').transpose()\n",
    "\n",
    "    # if no year is specified, return all years\n",
    "    elif country:\n",
    "      samples = pf6plus.loc[pf6plus.Country.isin([country])]\n",
    "      phenotypes = samples.groupby(['Year', drug]).size().unstack().fillna(0).astype(int)\n",
    "      print(drug + ' resistant samples in ' + str(country))\n",
    "    \n",
    "    # if no country is specified, return all years\n",
    "    else:\n",
    "      if bin:\n",
    "        samples = pf6plus.loc[pf6plus.Year.isin([b for b in range(min(year),max(year))])]\n",
    "        print(drug + ' resistant samples in all countries from ' + str(min(year)) + ' to ' + str(max(year)) )\n",
    "      \n",
    "      elif population:\n",
    "        samples = pf6plus.loc[(pf6plus.Population.isin([population]))]\n",
    "        phenotypes = samples.groupby(['Population', drug]).size().unstack().fillna(0).astype(int)\n",
    "        print(drug + ' resistant samples on all years')\n",
    "\n",
    "      else:\n",
    "        print(drug + ' resistant samples on all years')\n",
    "        phenotypes = pf6plus.groupby(['Country', drug]).size().unstack().fillna(0).astype(int)\n",
    "      \n",
    "    \n",
    "    phenotypes = phenotypes.assign(Total=phenotypes.sum(1))\n",
    "\n",
    "    # calculating the frequency using all samples, no matter how many they are! (note that some combinations will have a small number of samples, so the frequency will not be an adequeate estimate))\n",
    "    phenotypes['Resistant Frequency'] = [round(row[0]/(row[0]+row[1]), 2) for row in phenotypes[['Resistant','Sensitive']].to_numpy()]\n",
    "\n",
    "    # fully troubleshoot numbers (TO DO)\n",
    "    # add exception for multiple countries & multiple years (TO DO)\n",
    "    # add exception for repeated countries/years (TO DO)\n",
    "    # try:\n",
    "    #   return phenotypes['Resistant']\n",
    "    # except KeyError:\n",
    "    #   raise KeyError('The specified Country/Year combination is not in the dataset')\n",
    "\n",
    "    return(phenotypes)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "6QTu5VhHc7u8"
   },
   "source": [
    "###Â Plot Drug Resistant Prevalence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "Tzq95CrSc915"
   },
   "outputs": [],
   "source": [
    "def plot_dr_prevalence(drugs, country = None, population = None, year = None, bin=False):\n",
    "    \n",
    "    ''' Plot the prevalence of resistant samples per country/year \n",
    "    \n",
    "    Parameters:\n",
    "      - drug: Any/list of the drugs in the Pf6+ dataframe ['Artemisinin', 'Chloroquine', 'DHA-PPQ', 'Piperaquine', 'Pyrimethamine', 'S-P', 'S-P-IPTp', 'Sulfadoxine']\n",
    "      - country: Any of the countries in the Pf6+ dataframe (if specified, population value is not used) ['Bangladesh', 'Benin', 'Burkina Faso', 'Cambodia', 'Cameroon', 'Colombia', 'Congo DR', 'Ethiopia', 'Gambia', 'Ghana', 'Guinea', 'India', 'Indonesia', 'Ivory Coast', 'Kenya', 'Laos', 'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mozambique', 'Myanmar', 'Nigeria', 'Papua New Guinea', 'Peru', 'Senegal', 'Tanzania', 'Thailand', 'Uganda', 'Viet Nam']\n",
    "      - population: Any of the populations in the Pf6+ dataframe ['CAF', 'EAF', 'ESEA', 'OCE', 'SAM', 'SAS', 'WAF', 'WSEA']\n",
    "      - year: Any/list of the years in the Pf6+ dataframe [2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]\n",
    "      - bin: If True, all the years between the specified values will be used. If False, individual years are used.\n",
    "\n",
    "    Returns:\n",
    "      A series of plots (one per drug) showing the prevalence of resistant variants the drug/country/year (or) drug/country/year combination provided.\n",
    "      Note that to increase confidence on disperse data, we only use country (or) population/year combinations with at least 25 samples.\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    locus_year_group = pd.DataFrame(pf6plus.loc[(pf6plus['Country'] == country) & (pf6plus['Population'] == population)])\n",
    "    pf_tmp = pd.DataFrame(locus_year_group.groupby(['Year', 'Country']).size())\n",
    "    pf_tmp = pf_tmp.reset_index()\n",
    "    pf_country = pf6plus.loc[(pf6plus['Year'].isin(pf_tmp[pf_tmp[0]>25]['Year'])) & (pf6plus['Country'].isin(pf_tmp[pf_tmp[0]>25]['Country']))]\n",
    " \n",
    "    locus_year_group = pd.DataFrame(pf6plus.loc[(pf6plus['Population'] == population)])\n",
    "    pf_tmp = pd.DataFrame(locus_year_group.groupby(['Year', 'Country']).size())\n",
    "    pf_tmp = pf_tmp.reset_index()\n",
    "    samples_subset = pf6plus.loc[(pf6plus['Year'].isin(pf_tmp[pf_tmp[0]>25]['Year'])) & (pf6plus['Country'].isin(pf_tmp[pf_tmp[0]>25]['Country']))]\n",
    "    pf = samples_subset.loc[(samples_subset['Country'] != country) & (samples_subset['Population'] == population)]\n",
    "    \n",
    "    #define axs in fig based on # of drugs\n",
    "    fig_cols = len(drugs)\n",
    "\n",
    "    if fig_cols <= 6:\n",
    "      fig_rows = 2\n",
    "    else:\n",
    "      fig_rows = 3\n",
    "    \n",
    "\n",
    "    if fig_cols == 1:\n",
    "      fig, axes = plt.subplots(fig_cols, fig_rows, figsize=(2*7,fig_cols*4))\n",
    "    else:\n",
    "      fig, axes = plt.subplots(fig_cols, fig_rows, figsize=(fig_cols*7,fig_cols*4))\n",
    "\n",
    "    for i, ax in enumerate(axes.flatten()):\n",
    "\n",
    "        if (i<len(drugs)):\n",
    "            phenotypes = pf.groupby(['Year', drugs[i]]).size().unstack().fillna(0).astype(int)\n",
    "            phenotypes = phenotypes.assign(Total=phenotypes.sum(1))\n",
    "            \n",
    "            p = pf_country.groupby(['Year', drugs[i]]).size().unstack().fillna(0).astype(int)\n",
    "            p = p.assign(Total=p.sum(1))\n",
    "\n",
    "            #check if it is in the data before giving the size \n",
    "            try:\n",
    "              p['Resistant']\n",
    "              p['Sensitive']\n",
    "            except:\n",
    "              #add diff exceptions messages (dependending on resistant, sensitive, undetermined)\n",
    "              print('Not enough data on ' + drugs[i] + ' resistant variants in dataset')\n",
    "              fig.delaxes(axes.flatten()[i])\n",
    "              continue\n",
    "\n",
    "            #add exceptions for spellings and countries not matching the population provided (allow but throw a warning)\n",
    "            if ('Resistant' in phenotypes) & ('Sensitive' in phenotypes):\n",
    "                phenotypes['Prevalence ' + population] = [round(row[0]/(row[0]+row[1]), 2) for row in phenotypes[['Resistant','Sensitive']].to_numpy()]\n",
    "                p['Prevalence ' + country] = [round(row[0]/(row[0]+row[1]), 2) for row in p[['Resistant','Sensitive']].to_numpy()]\n",
    "\n",
    "            elif ('Resistant' in phenotypes ):\n",
    "                phenotypes['Prevalence ' + population] = [round(row[0]/(row[0]), 2) for row in phenotypes[['Resistant']].to_numpy()]\n",
    "                p['Prevalence ' + country] = [round(row[0]/(row[0]), 2) for row in p[['Resistant']].to_numpy()]\n",
    "                \n",
    "            elif ('Sensitive' in phenotypes):\n",
    "                phenotypes['Prevalence ' + population] = [round(row[0]/(row[0]), 2) for row in phenotypes[['Sensitive']].to_numpy()]\n",
    "                p['Prevalence ' + country] = [round(row[0]/(row[0]), 2) for row in p[['Sensitive']].to_numpy()]\n",
    "\n",
    "            ax.set_ylim(0,1)\n",
    "            ax.set_xlim(2001,2019)\n",
    "            \n",
    "            ax.plot(phenotypes['Prevalence ' + population],'o-')\n",
    "            ax.plot(p['Prevalence ' + country],'o-')\n",
    "\n",
    "            ax.set_title(drugs[i] +' resistance prevalence in ' + country + ' from ' + str(min(phenotypes.index)) + ' to '+ str(max(phenotypes.index)), size= 15)\n",
    "            ax.legend(['Prevalence ' + population, 'Prevalence ' + country])\n",
    "\n",
    "            #delete unused axis (need to reorder drugs before plotting, avoid whitespace)\n",
    "            axes = axes.flatten()\n",
    "            for j in range(fig_cols, fig_cols*fig_rows):\n",
    "                axes[j].axis(\"off\")\n",
    "            \n",
    "            #implement evenly spaced yearly bins TO DO\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "wFAgMEk0aHhL"
   },
   "source": [
    "### Plot most common haplotypes per population/country"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "rIB0Ck0yZnt8"
   },
   "outputs": [],
   "source": [
    "def plot_haplotype_frequency(gene, top_haplotypes=5, country = None, pop = None, year = None, bin=False):\n",
    "    \n",
    "    ''' Tabulate the frequency of top n haplotypes on a specife gene per country (or) population per year \n",
    "    \n",
    "    Parameters:\n",
    "      - gene: Any of the genes in the Pf6+ dataframe ['PfCRT', 'Kelch', 'PfDHFR', 'PfEXO', 'PGB', 'Plasmepsin2/3', 'PfDHPS', 'PfMDR1']\n",
    "      - top_haplotypes: The (n) most common haplotypes, default is 5. These excludes missing haplotypes.\n",
    "      - country: Any of the countries in the Pf6+ dataframe (if specified, population value is not used) ['Bangladesh', 'Benin', 'Burkina Faso', 'Cambodia', 'Cameroon', 'Colombia', 'Congo DR', 'Ethiopia', 'Gambia', 'Ghana', 'Guinea', 'India', 'Indonesia', 'Ivory Coast', 'Kenya', 'Laos', 'Madagascar', 'Malawi', 'Mali', 'Mauritania', 'Mozambique', 'Myanmar', 'Nigeria', 'Papua New Guinea', 'Peru', 'Senegal', 'Tanzania', 'Thailand', 'Uganda', 'Viet Nam']\n",
    "      - population: Any of the populations in the Pf6+ dataframe ['CAF', 'EAF', 'ESEA', 'OCE', 'SAM', 'SAS', 'WAF', 'WSEA']\n",
    "      - year: Any/list of the years in the Pf6+ dataframe [2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]\n",
    "      - bin: If True, all the years between the specified values will be used. If False, individual years are used.\n",
    "\n",
    "    Returns:\n",
    "      A dataframe showing the number of Resistant, Sensitive & Undetermined \n",
    "      samples using the drug/country/year (or) drug/country/year combination provided. . The total number\n",
    "      of samples and drug resistant frequency is also provided. \n",
    "    '''\n",
    "   \n",
    "    if pop:\n",
    "      loc = 'Population'\n",
    "      location = pop\n",
    "    elif country:\n",
    "      loc = 'Country'\n",
    "      location = country\n",
    "    else:\n",
    "      print('No country or population provided')\n",
    "\n",
    "\n",
    "    for population in location:\n",
    "        #get samples size for country selected\n",
    "        locus_year_group = pd.DataFrame(pf6plus.loc[(pf6plus[loc] == population)])\n",
    "\n",
    "        pf = pd.DataFrame(locus_year_group.groupby(['Year', 'Country']).size())\n",
    "        pf = pf.reset_index()\n",
    "\n",
    "        #get year/country combinations for each point selected\n",
    "        samples_subset = pf6plus.loc[(pf6plus['Year'].isin(pf[pf[0]>25]['Year'])) & (pf6plus['Country'].isin(pf[pf[0]>25]['Country']))]\n",
    "\n",
    "        top_haps = list(samples_subset.groupby(gene).size().sort_values(ascending=False).index)\n",
    "\n",
    "        # exclude missing values for all haplotypes\n",
    "        if (top_haps):\n",
    "            missing = '-'* len(top_haps[0])\n",
    "            \n",
    "            if (missing in top_haps):\n",
    "                top_haps.remove(missing)\n",
    "\n",
    "            # individual missing values for Kelch\n",
    "            if ('-' in top_haps):\n",
    "                top_haps.remove('-')\n",
    "            \n",
    "            for hap in top_haps[:]:\n",
    "               \n",
    "                if (',' in hap) | ('-' in hap): \n",
    "                    top_haps.remove(hap)\n",
    "\n",
    "        phenotypes = samples_subset.groupby(['Year', gene]).size().unstack().fillna(0).astype(int)\n",
    "\n",
    "        plt.xlim(2001,2019)\n",
    "        plt.ylim(0,1)\n",
    "        \n",
    "        phenotypes_norm=[]\n",
    "\n",
    "        for i in range(0,len(phenotypes)):\n",
    "            phenotypes_n=[]\n",
    "            for j in range(0,len(top_haps)):\n",
    "              phenotypes_n.append(phenotypes[top_haps[j]].values[i])\n",
    "\n",
    "            phenotypes_norm.append(sum(phenotypes_n))\n",
    "\n",
    "        if (len(top_haps)>=5):\n",
    "            ordered_haps = sorted([top_haps[0], top_haps[1], top_haps[2], top_haps[3], top_haps[4]])\n",
    "            \n",
    "            #check dynamic changes of colours (TO DO)\n",
    "            all_proportions = phenotypes[ordered_haps[0]] /(phenotypes_norm) + phenotypes[ordered_haps[1]] /(phenotypes_norm) + phenotypes[ordered_haps[2]] /(phenotypes_norm) + phenotypes[ordered_haps[3]] /(phenotypes_norm)+ phenotypes[ordered_haps[4]] /(phenotypes_norm)\n",
    "\n",
    "            plt.plot(phenotypes[ordered_haps[0]] /phenotypes_norm , label = ordered_haps[0], color= 'green',linestyle='--', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[1]] /phenotypes_norm , label = ordered_haps[1], color= 'blue',linestyle='--', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[2]] /phenotypes_norm , label = ordered_haps[2], color= 'magenta',linestyle='--', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[3]] /phenotypes_norm, label = ordered_haps[3], color= 'yellow',linestyle='--', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[4]] /phenotypes_norm , label = ordered_haps[4], color= 'orange',linestyle='--', marker='o')\n",
    "            plt.plot(1-all_proportions, label = \"Other Haplotypes\", color= 'grey',linestyle='--', marker='o')\n",
    "\n",
    "        elif (len(top_haps)==4):\n",
    "            ordered_haps = sorted([top_haps[0], top_haps[1], top_haps[2], top_haps[3]])\n",
    "            all_proportions = phenotypes[ordered_haps[0]] /(phenotypes_norm) + phenotypes[ordered_haps[1]] /(phenotypes_norm) + phenotypes[ordered_haps[2]] /(phenotypes_norm) + phenotypes[ordered_haps[3]] /(phenotypes_norm)\n",
    "\n",
    "            plt.plot(phenotypes[ordered_haps[0]] /phenotypes_norm , label = ordered_haps[0], color= 'green')\n",
    "            plt.plot(phenotypes[ordered_haps[1]] /phenotypes_norm , label = ordered_haps[1], color= 'blue')\n",
    "            plt.plot(phenotypes[ordered_haps[2]] /phenotypes_norm , label = ordered_haps[2], color= 'green')\n",
    "            plt.plot(phenotypes[ordered_haps[3]] /phenotypes_norm, label = ordered_haps[3], color= 'red')\n",
    "            plt.plot(1-all_proportions, label = \"Other Haplotypes\", color= 'grey',linestyle='--', marker='o')\n",
    "            plt.plot(1-all_proportions, label = \"Other Haplotypes\", color= 'grey',linestyle='--', marker='o')\n",
    "\n",
    "\n",
    "        elif (len(top_haps)==3):\n",
    "            ordered_haps = sorted([top_haps[0], top_haps[1], top_haps[2]])\n",
    "            all_proportions = phenotypes[ordered_haps[0]] /(phenotypes_norm) + phenotypes[ordered_haps[1]] /(phenotypes_norm) + phenotypes[ordered_haps[2]] /(phenotypes_norm) \n",
    "\n",
    "            plt.plot(phenotypes[ordered_haps[0]] /phenotypes_norm , label = ordered_haps[0], color= 'blue', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[1]] /phenotypes_norm , label = ordered_haps[1], color= 'yellow', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[2]] /phenotypes_norm , label = ordered_haps[2], color= 'green', marker='o')\n",
    "            plt.plot(1-all_proportions, label = \"Other Haplotypes\", color= 'grey',linestyle='--', marker='o')\n",
    "\n",
    "\n",
    "        elif (len(top_haps)==2):\n",
    "            ordered_haps = sorted([top_haps[0], top_haps[1]])\n",
    "            all_proportions = phenotypes[ordered_haps[0]] /(phenotypes_norm) + phenotypes[ordered_haps[1]] /(phenotypes_norm) \n",
    "\n",
    "            plt.plot(phenotypes[ordered_haps[0]] /phenotypes_norm , label = ordered_haps[0], color= 'blue', marker='o')\n",
    "            plt.plot(phenotypes[ordered_haps[1]] /phenotypes_norm , label = ordered_haps[1], color= 'purple', marker='o')\n",
    "            plt.plot(1-all_proportions, label = \"Other Haplotypes\", color= 'grey',linestyle='--', marker='o')\n",
    "\n",
    "        elif (len(top_haps)==1):\n",
    "            ordered_haps = [top_haps[0]]\n",
    "            all_proportions = phenotypes[ordered_haps[0]] /(phenotypes_norm) \n",
    "            \n",
    "            plt.plot(phenotypes[ordered_haps[0]] /phenotypes_norm , label = ordered_haps[0], color= 'blue', marker='o')\n",
    "            plt.plot(1-all_proportions, label = \"Other Haplotypes\", color= 'grey',linestyle='--', marker='o')\n",
    "\n",
    "        plt.axvline(2012,color='grey', linestyle='--', linewidth=1)\n",
    "        \n",
    "        #add year exceptions (TO DO)\n",
    "        #add multiple countries /pops exceptions (TO DO)\n",
    "        #ensure inconsistencies between E/WSEA in Thailand are taken into account (TO DO)\n",
    "        #if no arg for population, take the first ocurrence (note that this could be a problem with countries in multiple populations (eg. Thailand))\n",
    "        # if pop is None:\n",
    "        #     pop = pf6plus.loc[pf6plus['Country']==country]['Population'][0]\n",
    "        \n",
    "        if country:\n",
    "          plt.title(gene + ' Predominant Variants in '+ (country[0]))\n",
    "        else:\n",
    "          plt.title(gene + ' Predominant Variants in '+ population)\n",
    "          \n",
    "        plt.legend()\n",
    "        plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "functions_270921.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
